src/Corax/IndexWriter.cs

```diff
@@ -29,14 +30,106 @@ public enum TermIdMask : long
         Set = 2
     }
 
+
+    public class IndexFieldsMapping : IEnumerable<IndexFieldBinding>
+    {
+        public static readonly IndexFieldsMapping Instance = new IndexFieldsMapping(null);
+
+        private readonly ByteStringContext _context;
+        private readonly Dictionary<Slice, IndexFieldBinding> _fields;
+        private readonly Dictionary<int, IndexFieldBinding> _fieldsById;
+        private readonly List<IndexFieldBinding> _fieldsList;
+
+        public int Count => _fieldsById.Count;
+
+        public IndexFieldsMapping(ByteStringContext context)        
+        {
+            _context = context;
+            _fields = new Dictionary<Slice, IndexFieldBinding>();
+            _fieldsById = new Dictionary<int, IndexFieldBinding>();
+            _fieldsList = new List<IndexFieldBinding>();
+        }
+
+        public IndexFieldsMapping AddBinding(int fieldId, Slice fieldName, Analyzer analyzer = null)
+        {
+            if (!_fieldsById.TryGetValue(fieldId, out var storedAnalyzer))
+            {
+                var binding = new IndexFieldBinding(fieldId, fieldName, analyzer);
+                _fields[fieldName] = binding;
+                _fieldsById[fieldId] = binding;
+                _fieldsList.Add(binding);
+            }
+            else
+            {
+                Debug.Assert(analyzer == storedAnalyzer.Analyzer);
```

Comment:
Why are we allowing the same thing twice?

Reply:
We are adding several different ways to access them because we need to serve different access patterns. The tradeoff is objects vs optimized access pattern. For example, enumerations are done on the list instead of a dictionary (which is far more expensive).

Reply:
Are we actually expecting to have duplicate calls, however?

Reply:
Yes. In fact this is a candidate class for further optimizations down the line using TRIE trees or something like that.